### DISCUSSION TITLE
Pitch detection for a guitar tuner

### DISCUSSION THREAD

**Discussion by xconverge** [cycfi/q#76] [General]
Thanks a lot for providing the library! I am using this on a stm32 (daisy seed) as a guitar tuner.

I have everything running great, but wanted to float some ideas on how to stabilize the readings a bit

I have a small frequency detection class (I wrapped the pitch detector with this because I was exploring some other pitch detection algorithms at the same time. It looks something like this:

```

float sample_rate = 48000;

// The frequency detection bounds;
frequency lowest_frequency = pitch_names::C[2];
frequency highest_frequency = pitch_names::C[5];

// The pitch detector
pitch_detector pd{lowest_frequency, highest_frequency, sample_rate, -40_dB};

// The pitch detector pre-processor
signal_conditioner::config preprocessor_config;
signal_conditioner preprocessor{preprocessor_config, lowest_frequency,
                                highest_frequency, sample_rate};

void FrequencyDetectorQ::Process(float in) {
  // Pre-process the signal for pitch detection
  float pd_sig = preprocessor(in);

  // send the processed sample through the pitch detector
  if (pd(pd_sig)) {
    m_cachedFrequency = as_float(frequency{pd.get_frequency()});
  }
}
```

It seems to work very well, is very responsive and accurate. The next thing I wanted to approach is how to stabilize it a bit to make it a bit easier to tune accurately. I tried both a moving average(and exp moving average) along with a LPF with a cutoff of 4 hz and they each seemed to minorly improve the ease of tuning. I applied these filters to the m_cachedFrequency essentially, the OUTPUT of the pitch detector.

Do you have any advice on algorithms to look into, where in the chain I should be looking (at the input signal/the preprocessor parameteres/the output final frequency)?

Also I am eagerly waiting to check out Hz when you release it after reading your blog entries, thanks again!

**Comment by djowel** [comment]
I'd use the [Dynamic Smoother](https://cycfi.github.io/q/q/v1.5-dev/reference/misc/dynamic_smoother.html) with a very low cutoff. I'd even try to go as low as 0.1Hz, then slowly increase for faster response.

**Reply by xconverge** [reply]
On the audio signal or on the detected frequency? With that cutoff it must be the detected frequency right?

**Reply by djowel** [reply]
detected frequency

**Reply by xconverge** [reply]
Thanks @djowel this is definitely on the right track, I think the problem I am having before/still is how to set the cutoff specifically for use with the pitch detector due to the way the pitch detector has the "ready" state.

My sample rate is 48000, but the frequency is only getting changed whenever the pitch detector is reporting it is ready, so to me, this seems like it is going to mess up the math?

```
  // Pre-process the signal for pitch detection
  float preProcessedSignal = m_preProcessor->operator()(in);

  // Send the processed sample through the pitch detector
  const bool ready = m_pitchDetector->operator()(preProcessedSignal);

  // If result is ready, get the detected frequency
  if (ready) {
    m_currentFrequency = m_pitchDetector->get_frequency();

    // Run a smoothing filter on the detected frequency
    m_currentFrequency = m_smoothingFilter->operator()(m_currentFrequency);
  }
```

Moving the smoothing filter outside of the ready condition doesn't seem to make sense to me either. Because of all of this, when I set the sample rate to 48000 for the filter, I need to lie to it and give it a cutoff of ~20 or so hz. I also don't think it is a good idea to ignore this ready state? (but you would know best of course)

This is why I started with a moving average and am a little lost on the best way to proceed

**Reply by djowel** [reply]
I'd add an else clause and I'd call the `m_smoothingFilter` with a frequency of zero. Something like this:

```
  if (ready) {
    m_currentFrequency = m_pitchDetector->get_frequency();

    // Run a smoothing filter on the detected frequency
    m_currentFrequency = m_smoothingFilter->operator()(m_currentFrequency);
  }
  else {
     m_currentFrequency = m_smoothingFilter->operator()(0);
  }
```

Note: this is not quite optimal. I can probably add a quick way to set the dynamic_smoother's value to an arbitrary value. Some filters already allow you to do that using the operator=.

**Reply by djowel** [reply]
BTW,  I'd call m_smoothingFilter like this:

```
(*m_smoothingFilter)(m_currentFrequency);
```

less verbosity. But of course, it's just my style.

**Reply by djowel** [reply]
BTW, there's also the [1€ Filter](https://gery.casiez.net/1euro/) that I've been wanting to add to Q. It's also designed to minimize jitter and lag in the human response range. Unlike fixed sampling rates, the 1€ filter uses a time parameter that specifies the elapsed time since the previous reading.

Here's a simple implementation in python: https://github.com/jaantollander/OneEuroFilter

**Reply by xconverge** [reply]
I will try it out tomorrow, thanks!

**Reply by djowel** [reply]
I added operator= to `dynamic_smoother`. You can use this code:

```
  if (ready) {
    m_currentFrequency = m_pitchDetector->get_frequency();

    // Run a smoothing filter on the detected frequency
    m_currentFrequency = m_smoothingFilter->operator()(m_currentFrequency);
  }
  else {
     *m_smoothingFilter = 0;
     m_currentFrequency =  0;
  }
```

**Reply by xconverge** [reply]
I need to keep the latest non-zero value, m_currentFrequency is the variable I am displaying and using for the rest of my "tuning". I would probably do something like this instead:

```
  if (ready) {
    m_currentFrequency = m_pitchDetector->get_frequency();

    // Run a smoothing filter on the detected frequency
    m_currentFrequency = m_smoothingFilter->operator()(m_currentFrequency);
  }
  else {
     *m_smoothingFilter = m_currentFrequency;
  }
```

I think a time delta based version of a filter probably makes the most sense in this case.

**Reply by djowel** [reply]
> I need to keep the latest non-zero value, m_currentFrequency is the variable I am displaying and using for the rest of my "tuning". I would probably do something like this instead:
> 
> ```
>   if (ready) {
>     m_currentFrequency = m_pitchDetector->get_frequency();
> 
>     // Run a smoothing filter on the detected frequency
>     m_currentFrequency = m_smoothingFilter->operator()(m_currentFrequency);
>   }
>   else {
>      *m_smoothingFilter = m_currentFrequency;
>   }
> ```
> 
> I think a time delta based version of a filter probably makes the most sense in this case.

In this case, just ignore the else clause. The filter will do just fine since it caches the latest value. It will be, as-if, it is held in suspension, until brought back to life again. The math would still be the same (well, almost, but the difference is negligible in your use-case).

**Reply by xconverge** [reply]
The 1 euro filter is looking really good so far! Using this one https://github.com/casiez/OneEuroFilter/blob/main/CppUsingTemplates/1efilter.cc

**Reply by djowel** [reply]
> The 1 euro filter is looking really good so far! Using this one https://github.com/casiez/OneEuroFilter/blob/main/CppUsingTemplates/1efilter.cc

Wonderful! Does it work better than `dynamic_smoother`?

**Reply by xconverge** [reply]
I need to tune and play with the beta value a bit more, but with my initial values I think it seems a bit better. It just feels nice to use a time delta (would be straightforward to add to any regular lpf implementation of course instead of a fixed sample rate) when I don't really know the sample rate of the pitch detector being ready/outputting a result

**Reply by djowel** [reply]
> It just feels nice to use a time delta

Agreed. That is one of its appeals that got me interested in adding it to Q DSP. We all need parameter jitter smoothing somehow.

**Comment by djowel** [comment]
Unrelated question: Why hold the filter(s) by pointer? These objects are very small and efficient and it is quite optimal to hold them by value.

**Reply by xconverge** [reply]
I had it by value initially, and certainly would prefer it in this case as well, the "problem" I ran into was that I only had/knew the sample rate at a later time post-construction with the framework I was/am working with. I had hardcoded my sample rate when I was using it initially holding it by value.

If the filter had an init() function (that took in a sample rate at minimum, all parameters would be nice too though) I would have been in good shape as well

https://github.com/bkshepherd/DaisySeedProjects/pull/21

**Reply by djowel** [reply]
Aww, just assign by value :-)

```
m_smoothingFilter = q::dynamic_smoother{base, sps};
```

`init` is sooo 90's :-)

**Reply by xconverge** [reply]
That works for the dynamic_smoother but what about the pitch_detector and signal_conditioner, I was trying to stay consistent

**Reply by djowel** [reply]
Should be the same. All objects in Q DSP can be constructed and assigned. But, hey, I do not want to be pedantic. If it works for you, fine :-)

**Reply by xconverge** [reply]
```
  cycfi::q::pitch_detector test;
  test = cycfi::q::pitch_detector{lowest_frequency, highest_frequency,
                                  sample_rate, lin_to_db(0)};
```

```
error: use of deleted function 'cycfi::q::pitch_detector& cycfi::q::pitch_detector::operator=(const cycfi::q::pitch_detector&)'
   60 |                                   sample_rate, lin_to_db(0)};
```

**Reply by djowel** [reply]
> ```
>   cycfi::q::pitch_detector test;
>   test = cycfi::q::pitch_detector{lowest_frequency, highest_frequency,
>                                   sample_rate, lin_to_db(0)};
> ```
> 
> ```
> error: use of deleted function 'cycfi::q::pitch_detector& cycfi::q::pitch_detector::operator=(const cycfi::q::pitch_detector&)'
>    60 |                                   sample_rate, lin_to_db(0)};
> ```

It's a bug :-( Easy to fix. Can you file a separate issue?

**Comment by xconverge** [comment]
Now that I have some of those details sorted out, the only minor quirk I notice is when switching between the low E and high e string, the frequency detector seems to be a bit sticky towards the lower frequency, potentially some sort of harmonic?

If I move through the strings in series tuning E A D G B e, it seems to be a bit more reliable/quick to detect each specific string vs if I tune E and then e back to back. I need to pluck the high e string a bit louder or with more intention to have it register in the ~329 hz range instead of the 83 hz range of the low e string

Could also be sympathetic resonance

**Reply by djowel** [reply]
> Now that I have some of those details sorted out, the only minor quirk I notice is when switching between the low E and high e string, the frequency detector seems to be a bit sticky towards the lower frequency, potentially some sort of harmonic?
> 
> If I move through the strings in series tuning E A D G B e, it seems to be a bit more reliable/quick to detect each specific string vs if I tune E and then e back to back. I need to pluck the high e string a bit louder or with more intention to have it register in the ~329 hz range instead of the 83 hz range of the low e string
> 
> Could also be sympathetic resonance

Could be. Share a recording of this (wav file preferably) and I can try it out with the new Hz PD. I'm not working on the older algorithm anymore.

**Reply by xconverge** [reply]
I think a .wav makes a lot of sense and probably overdue instead of being hunched over my guitar/amp/pedals with a laptop on the floor! I will get one in a few hours. I think this has all piqued my interest enough to write some test scripts and really be able to tune things.

Thanks for the dialog, I really appreciate it and the opportunity to learn!

**Reply by xconverge** [reply]
Might not be a perfect clip, but it's a start

[test.txt](https://github.com/user-attachments/files/17536991/test.txt) rename to .wav

**Reply by xconverge** [reply]
This is roughly where I am starting, nothing looks too crazy, pretty tough without some more normalization potentially, pretty uneven playing by me and a missing "G" detection around the 8 second mark, not sure...

```cpp
#include <cmath>
#include <fstream>
#include <iostream>
#include <q/fx/signal_conditioner.hpp>
#include <q/pitch/pitch_detector.hpp>
#include <q/support/literals.hpp>
#include <q/support/pitch_names.hpp>
#include <q_io/audio_file.hpp>
#include <vector>

using namespace cycfi::q;

template <typename T = double>
struct low_pass_filter {
  low_pass_filter() : hatxprev(0), xprev(0), hadprev(false) {}
  T operator()(T x, T alpha) {
    T hatx;
    if (hadprev) {
      hatx = alpha * x + (1 - alpha) * hatxprev;
    } else {
      hatx = x;
      hadprev = true;
    }
    hatxprev = hatx;
    xprev = x;
    return hatx;
  }
  T hatxprev;
  T xprev;
  bool hadprev;
};

template <typename T = double, typename timestamp_t = double>
struct one_euro_filter {
  one_euro_filter(double _freq, T _mincutoff, T _beta, T _dcutoff)
      : freq(_freq),
        mincutoff(_mincutoff),
        beta(_beta),
        dcutoff(_dcutoff),
        last_time_(-1) {}
  T operator()(T x, timestamp_t t = -1) {
    T dx = 0;

    if (last_time_ != -1 && t != -1 && t != last_time_ && t > last_time_) {
      freq = 1.0 / (t - last_time_);
    }
    last_time_ = t;

    if (xfilt_.hadprev) dx = (x - xfilt_.hatxprev) * freq;

    T edx = dxfilt_(dx, alpha(dcutoff));
    T cutoff = mincutoff + beta * std::abs(static_cast<double>(edx));
    return xfilt_(x, alpha(cutoff));
  }

  double freq;
  T mincutoff, beta, dcutoff;

 private:
  T alpha(T cutoff) {
    T tau = 1.0 / (2 * M_PI * cutoff);
    T te = 1.0 / freq;
    return 1.0 / (1.0 + tau / te);
  }

  timestamp_t last_time_;
  low_pass_filter<T> xfilt_, dxfilt_;
};

// Inputs:
// Estimated frequency: Overwritten by timestamps at runtime and not used
// Cutoff Freq
// Beta: 0.0f disables it entirely, but used for scaling cutoff frequency
// Derivative cutoff freq: used when beta is > 0
one_euro_filter<float, float> smoothingFilter{48000, 0.5f, 0.05f, 1.0f};

int main() {
  wav_reader src{"test/audio_files/test.wav"};
  float const sps = src.sps();

  std::vector<float> input(src.length());
  src.read(input);

  std::cout << "Starting:" << std::endl;

  std::cout << "Num Samples: " << src.length() << std::endl;
  std::cout << "Sample Rate: " << sps << std::endl;
  std::cout << "Length (s): " << src.length() / sps << std::endl;

  // The frequency detection bounds;
  frequency lowest_frequency = pitch_names::C[2];
  frequency highest_frequency = pitch_names::C[5];

  cycfi::q::pitch_detector pd{lowest_frequency, highest_frequency, sps,
                              lin_to_db(0)};

  signal_conditioner::config preprocessor_config;
  signal_conditioner preprocessor{preprocessor_config, lowest_frequency,
                                  highest_frequency, sps};

  std::ofstream csv("test/results/sk_output.csv");

  for (int i = 0; i < input.size(); i++) {
    // Small amount of input gain due to recording quality probably
    const float in = input[i] * 1.5f;

    // Pre-process the signal for pitch detection
    float preProcessedSignal = preprocessor(in);

    // Send the processed sample through the pitch detector
    const bool ready = pd(preProcessedSignal);

    // If result is ready, get the detected frequency
    if (ready) {
      const float freqRaw = pd.get_frequency();

      const float timestampSeconds = i / 48000.0f;

      const float freqFiltered = smoothingFilter(freqRaw, timestampSeconds);

      csv << timestampSeconds << "," << freqRaw << "," << freqFiltered
          << std::endl;
    }
  }

  std::cout << "Done" << std::endl;
}

```

![Figure_1](https://github.com/user-attachments/assets/51b13102-27c2-4a47-9ec9-2bc864ec5b40)

**Reply by xconverge** [reply]
As a side note, the 1 euro filter seems to be doing pretty well for the limited tuning I did

![2](https://github.com/user-attachments/assets/6111ee38-a8a0-43f4-a13e-324f145fcb39)

**Reply by djowel** [reply]
> Might not be a perfect clip, but it's a start
> 
> [test.txt](https://github.com/user-attachments/files/17536991/test.txt) rename to .wav

You'll probably need to optimize your gain stage. It's way too low at -15dB peak:

<img width="1116" alt="image" src="https://github.com/user-attachments/assets/527cc7bd-1205-48e1-be3a-a2aea3f6f2a6">

**Reply by xconverge** [reply]
I didn't generate the recording on my MCU/hardware, just with the same guitar/pickups. I will add some logging to my hardware to see if I can match the gain a bit better and at least confirm the input levels are decent (unlike this .wav). I wanted to generate the wav quick and dirty so I could write my test/visualization code for starters

Also this is why in my test script I put 

```
    // Small amount of input gain due to recording quality probably
    const float in = input[i] * 1.5f;
```

**Reply by djowel** [reply]
The Hz pitch tracker detects the pitches perfectly, including the harmonics at the trailing edge of some samples (listen carefully at the 5th and 7th notes).

![image](https://github.com/user-attachments/assets/1b03f128-011e-4cc3-a688-cdf5e9c0802b)

Well, almost. There's an eager (zero latency) prediction at the onset of the 6th note (note to self: need to investigate), but ever so briefly, and settles very quickly. Your tuner should not be affected.

**Reply by djowel** [reply]
> I didn't generate the recording on my MCU/hardware, just with the same guitar/pickups. I will add some logging to my hardware to see if I can match the gain a bit better and at least confirm the input levels are decent (unlike this .wav). I wanted to generate the wav quick and dirty so I could write my test/visualization code for starters
> 
> Also this is why in my test script I put
> 
> ```
>     // Small amount of input gain due to recording quality probably
>     const float in = input[i] * 1.5f;
> ```

Even then, the new Hz PD got it right, right out of the box. I didn't do anything to the signal.

**Reply by xconverge** [reply]
Well that is fun news :)

**Reply by djowel** [reply]
> Well that is fun news :)

Fun indeed :-) I added your wav to the test suite.

**Reply by djowel** [reply]
Oh, wait, it gated some signals! I'll need to adjust the gain a bit.

**Reply by djowel** [reply]
> Oh, wait, it gated some signals! I'll need to adjust the gain a bit.

Here's the updated result:

![image](https://github.com/user-attachments/assets/ed044364-bf47-4928-ad9b-3a6b4db22c9c)

Note: The signal is weak and noisy. I hear hum. Listen carefully. You might want to add an AGC and filter the power supply a bit more. Or you might want to add a hum filter if this is unavoidable.

**Reply by djowel** [reply]
![image](https://github.com/user-attachments/assets/42faccc2-c16f-4dc1-a4bf-b4787971c415)

These are due to weak signal and hum. Listen carefully.

**Reply by xconverge** [reply]
I will add an agc or some sort of normalization on the input to all of this, I think that makes a lot of sense of course! I assumed the signal_conditioner was doing this somewhat, but maybe not with the default config parameters I am using? Or was I getting gated in the conditioner due to low signal level before any makeup gain was applied?

**Reply by djowel** [reply]
> I will add an agc or some sort of normalization on the input to all of this, I think that makes a lot of sense of course! I assumed the signal_conditioner was doing this somewhat, but maybe not with the default config parameters I am using? Or was I getting gated in the conditioner due to low signal level before any makeup gain was applied?

Oh, I'm sorry, I might not have been clear. I meant on the analog side, before going to the MCU. You'd want the MCU to receive the full signal swing as much as possible. Actually, for the purpose of tuning, it's even OK to have some saturation on the peaks. And the hum, that's something that you'd want to take care of—it's quite audible. A 60 Hz signal on top of with the main signal will confuse the pitch detector because you have two frequencies instead of a monophonic signal.

**Reply by xconverge** [reply]
Interesting, my other guitar doesn't have the same hum, even when searching for it. I also rarely play clean, I nearly always have distortion so the hum is not an issue with that guitar outside of signal processing for pitch detection 😆 The pickup in the guitar I used in that original recording is very high output (humbucker as well though) so I think it just comes with the territory for that guitar. Tuners can't be too picky with the input signals going in since there are so many variations of pickups (active and passive as well). The results and accuracy with the non-humming guitar might seem marginally better, but I also have confirmation bias for this to be the case a bit now that I know it is a cleaner signal.

I wanted to share a demo of the "finished" result, to show you how usable it is. This is potentially the least exciting guitar related youtube video that could be made, but I figured you would appreciate seeing it all come together after the conversation!

https://www.youtube.com/watch?v=FQ_RqBbvLtw

I don't think I really need to do much more tuning/adjustment and will just swap in hz when you release it :)

**Reply by djowel** [reply]
Well, I'm just a perfectionist. It probably doesn't matter in your case because you essentially average the results anyway, the fact that you are using a post smoothing filter. At the very least, increase the gain. But you might want to check for ground loops —the hum might be simply the way you recorded it.

**Reply by djowel** [reply]
> I wanted to share a demo of the "finished" result, to show you how usable it is. This is potentially the least exciting guitar related youtube video that could be made, but I figured you would appreciate seeing it all come together after the conversation!
> 
> https://www.youtube.com/watch?v=FQ_RqBbvLtw

Awesome! I'm collecting links right now of Q DSP applications. Do you have a github page, Instructables, or some such?

**Reply by djowel** [reply]
This one BTW: https://www.instructables.com/Making-a-Guitar-Tuner-and-Audio-Spectrum-Analyser-/

Is inspired by and uses BACF (his own implementation of).

**Reply by xconverge** [reply]
https://github.com/bkshepherd/DaisySeedProjects/pull/21 not sure if we will merge it due to pulling in additional dependencies but it is a really cool project

**Reply by djowel** [reply]
> [bkshepherd/DaisySeedProjects#21](https://github.com/bkshepherd/DaisySeedProjects/pull/21) not sure if we will merge it due to pulling in additional dependencies but it is a really cool project

Sweet!

**Reply by djowel** [reply]
"I explored quite a few ways of doing pitch detection (yin algorithm, basic period detection and zero crossing logic, etc) and joel over at cycfi has just knocked it out of the park with this one. I am eager for his updated library to test it out"

:-) I'm happy to hear the Q PD is working well for you! Could you give me a quick summary of why YIN didn't work out for you?

**Reply by xconverge** [reply]
> "I explored quite a few ways of doing pitch detection (yin algorithm, basic period detection and zero crossing logic, etc) and joel over at cycfi has just knocked it out of the park with this one. I am eager for his updated library to test it out"
> 
> :-) I'm happy to hear the Q PD is working well for you! Could you give me a quick summary of why YIN didn't work out for you?

I need to revisit it but to be honest I was battling a lot of variables at the same time. Now that things are stabilized and working, reworking some of my previous tests would be a good idea. I didn't give it a fair chance in the fight!

**Reply by xconverge** [reply]
I took another look this morning using https://github.com/ashokfernandez/Yin-Pitch-Tracking and tweaking it to take in floats

https://github.com/xconverge/DaisySeedProjects/compare/tweaks...xconverge:DaisySeedProjects:yin

Basically the detection is just SUPER jumpy. I could potentially tune the smoothing to make it work but I am happy with using q for now! I tried also using the q signal_condtioner as the input to yin as well to compare apples to apples a bit better, but no luck. I do see values that are sensible and accurate and speedy, there is just a LOT of noise around them.

I also could tune the buffer size and threshold a bit to improve it probably but 🤷

**Reply by xconverge** [reply]
Ok you got me there :)

https://github.com/bkshepherd/DaisySeedProjects/pull/25

I am decoupling my tuner implementation from the frequency detection with a class so that I can directly compare both back to back. I am going to continue with your implementation but Yin does seem to work quite well after some minor adjustments!

**Reply by djowel** [reply]
> Ok you got me there :)
> 
> [bkshepherd/DaisySeedProjects#25](https://github.com/bkshepherd/DaisySeedProjects/pull/25)
> 
> I am decoupling my tuner implementation from the frequency detection with a class so that I can directly compare both back to back. I am going to continue with your implementation but Yin does seem to work quite well after some minor adjustments!

Isn't YIN also a dependency? FWIW, Q-lib is header only. You could just as well take just the header files and use those as-is.

**Reply by xconverge** [reply]
The key word is "external". For Q-lib, the 2 "issues" that make it a more serious dependency are that I would need to change from c++11 to c++20, and the way the infra headers work. I do know that I could take some amount of headers and get there eventually. I do not consider a self contained single header implementation of yin to be an external dependency.

I plan on rebasing my PR using q and making it an option (and using it personally)

I also plan on pushing to change to c++20, but decoupling all of it makes it a bit easier to comprehend changesets IMO

**Reply by djowel** [reply]
> The key word is "external". For Q-lib, the 2 "issues" that make it a more serious dependency are that I would need to change from c++11 to c++20, and the way the infra headers work. I do know that I could take some amount of headers and get there eventually. I do not consider a self contained single header implementation of yin to be an external dependency.
> 
> I plan on rebasing my PR using q and making it an option (and using it personally)
> 
> I also plan on pushing to change to c++20, but decoupling all of it makes it a bit easier to comprehend changesets IMO

OK, that makes sense.

**Reply by xconverge** [reply]
I haven't fully benchmarked it fully yet, but the memory footprint and processing time of the implementation of yin I have seems non-ideal. I was spoiled by the performance and implementation ease using yours. 

Overall it works "fine" but if I want to passthrough my audio signal (unmuted tuner essentially), the processing time of yin every X amount of ticks gives me artifacts. I had no such issues using pitch_detector!

Anyways, thanks for the dialog, I have enjoyed it!

**Reply by djowel** [reply]
> Anyways, thanks for the dialog, I have enjoyed it!

As did I. Great to have company!

**Reply by xconverge** [reply]
We merged the q version and I deleted the yin version, so that project is certainly using your library now :)

**Comment by joulupukki** [comment]
@xconverge this is a very interesting discussion. I started a similar project using ESP32 and an external screen a few months back. It’d be interesting to pick your brain and swap ideas if you’re up to it. I’m planning to convert it over to a larger sceeen this weekend but am not fully satisfied with the smoothing yet. 

I’m using ESP-IDF. I believe I’m also using a 1EU filter. I found it to perform the best. But, I’ve never coded with LVGL before so the screen graphics are poor and I haven’t spent too much time on it once I saw Joel would be releasing a Hz lib.

My project is here: https://github.com/joulupukki/esp32-chromatic-tuner

Are you planning to publish your code?

**Reply by xconverge** [reply]
I still have my polytune on my board, and don't plan on getting rid of it, so I decided that I needed to get back to playing guitar instead of playing with my computer. I deemed it good enough once I merged it to the project and made it available to others :)

A few notes from reviewing your project:

I used the 1e filter primarily because I didn't want to hardcode a frequency (your project does). If you supply a time component to the 1e filter, it throws away this assigned frequency, and uses the supplied time since the previous sample to compute the frequency. I didn't want to hardcode a frequency because the pitch detection .ready() doesn't necessarily occur at a known frequency/rate. If this is the case, then tuning it can be a bit tough because setting 0.5 for a cutoff frequency isnt actually 0.5 hz, it is some alternate frequency scaled in some units you don't really know.

The code I did write is here:

https://github.com/bkshepherd/DaisySeedProjects/tree/main/Software/GuitarPedal

relevant files:
```
Util/frequency_detector_q.cpp
Util/frequency_detector_q.h
Util/1efilter.hpp
Effect-Modules/tuner_module.cpp
Effect-Modules/tuner_module.h
```

Joel did the hard part already with this library, the rest is really up to your specific needs. I didn't spend too much time tuning the 1e filter, but was pretty happy with the results from the small amount of time that I did spend.

I am pretty happy with the display logic I wrote in https://github.com/bkshepherd/DaisySeedProjects/blob/main/Software/GuitarPedal/Effect-Modules/tuner_module.cpp#L72 I think I did some simple but clever things by determining if I was sharp or flat, and by how much, then iterating forward or reverse from the middle through my `_ _ _ _ _ _ | _ _ _ _ _ _ _` display to decide if a block should be "active" or not.

I wrote the frequency detector with an interface originally so that I could substitute in different algorithms (and potentially Hz), but removed the interface once I settled on this library. I will update it and play with it in the future when hz is out.

Definitely feel free to ask any more specific questions if you have them, but your project is looking good!

**Reply by xconverge** [reply]
Yea I can't figure out how your 1e filter is getting a frequency... setting freq to 0 in the constructor, and then not providing a timestamp, seems to me like freq never gets tweaked and stays at 0, but that doesnt seem right because I would guess a divide by 0 or something funky would happen. I certainly could be missing something but I think passing a timestamp to .filter() would help make your tuning of the filter go better

**Reply by joulupukki** [reply]
Thanks @xconverge for the compliments. Clever display logic in yours! I've paused any improvements to my graphics output until I can solve the tuning.

I thought I'd switch to a larger screen, one that comes with the ESP32-ESP32-2432S028R (people refer to this as the Cheap Yellow Display / CYD: https://github.com/witnessmenow/ESP32-Cheap-Yellow-Display) because you can get them relatively easily and cheap and it's got a bigger screen. But, now that I've finally got it building with ESP-IDF, the frequency detection is completely off. I've tried changing the ADC sample rate to less than 20kHz but then it complains that it's out of range.

My latest branch is here: https://github.com/joulupukki/esp32-chromatic-tuner/tree/boyd/switch-to-cyd

I took all signal conditioning out and 1EU filter (temporarily – it's enabled in the code now) but ... something I'm doing isn't working correctly with this differen ESP32 chip.

The general idea in my project is that it reads from the ADC pin voltage values. With a friend's help, we've created external hardware that does some simple filtering of the signal and also does an offset voltage so that no signal is sitting about 1.45VDC. The ESP32 can read 3.3V or so and so when a string pluck happens, I see values between 0 and 4096. From that, I normalize the incoming values so that they are values between -1 and +1 and use those values to send into Q DSP's pitch detector.

To detect a signal, I'm ignoring the input when the difference between the low and high voltages is less than 50mV. That mostly worked in the older code and seems to kinda work still now.

In any case, I know a lot of the reason why I don't have this new code working is that there's gotta be stuff about sampling that I don't quite understand correctly. I've also heard that ESP32's ADC is notoriously bad and maybe in this particular ESP32 it's even worse than the ESP32-S3 (Heltec brand) I was using before?

**Reply by joulupukki** [reply]
I think another problem I've got is that when I turn the LCD on (attached to this ESP32) the noise it generates (and gets picked up by the hardware preamp) is insane and is probably throwing off my frequency measurement. I have an oscilloscope hooked up to the output of the preamp (uses an LM358 opamp) and it looks good until I turn the LCD on. Even with the LCD off though the frequency detection is not working. I'm guessing it has something to do with the sample rate I'm using, but I've tried all sorts of variations.

I'm assuming that the Frequency Rate that I sample with is the same rate that I should pass into Q DSP's pitch detector. I've tried 20kHz and am currently debugging with 48kHz, which is very much overkill, but I'm just trying to get _something_ to work. :)

Edit: Maybe false alarm on the LCD causing much interference. I just tried again ... maybe I had a wire loose or something but at least I'm not seeing crazy noise on the output signal of the op amp anymore when the LCD is powered on.

**Reply by xconverge** [reply]
Do you have a way of viewing the waveform you are trying to analyze? If not I would probably write some code that does some sort of rolling buffer that allows you to dump a sample to be looked at/processed on your computer. A .wav of hitting one or two strings would go a long way in decoupling your hw/sample rate/input side of things from your pitch detection/signal processing side of things. That is how I would approach this!

**Reply by joulupukki** [reply]
Yeah, I recorded a pure sine wave into a looper pedal and can play that back and I’ve analyzed it with my oscilloscope on the output of the preamp and that’s all good. I think there’s something I’m not quite understanding yet with how ADC sampling works on ESP32. I’m still digging to see if I can glean more info so that it’s not just magic.

**Reply by xconverge** [reply]
Oops sorry I didn't even see the second reply from you! My bad!

**Reply by djowel** [reply]
Have you considered using a dev board that is designed for audio, such as the Teensy or Daisy? Mixed signal electronics is difficult. I'm not sure if the ESP32 is even designed to take in audio inputs. Keep in mind that even at 12-bits (4096 steps), at 3.3v, you are dealing with an LSB of 80uV! I seriously doubt if the ESP32 analog electronics is that clean.

**Reply by joulupukki** [reply]
Yes. I’ve considered trying out one of the other boards, specifically the Daisy. However, part of my motivation on this is to figure out if there’s a fairly inexpensive way to make something decently functional. I had it mostly working with a different ESP32 that had a smaller screen (made by HELTEC), so I am still hopeful to make this work.

I’m not sure what “LSB” means (except for in ham radio terms - lower side band).

I am getting consistent readings right now, but feeding a 440Hz clean sine wave into it results in 42.5Hz, completely off. I’m going to try re-reading the ESP-IDF ADC continuous read API documentation again to see if there’s something I’m not understanding. I assume that I need to be passing in the read frequency into the pitch detection (sample rate)?

My hunch right now is that it has something to do with the buffer pool size and the number of frames I’m passing into the ADC continuous read API. We’ll see. I’m not quite ready to give up entirely. It seems like in this case (for audio) that the audio signal doesn’t have to be pristine (since I’m not processing the actual audio and outputting some effect) so hopefully I’ll be able to get something working soon. If, after lots more exploring, it fails, I’ll probably bag the project or just move to a Daisy Seed – or follow @xconverge’s path and buy a commercial offering and just practice my guitar playing more. ;)

**Reply by xconverge** [reply]
I highly recommend trying the daisy seed or a teensy, as part of the exploration! I really enjoyed being able to focus on the application level instead of the hardware level for this project

**Reply by joulupukki** [reply]
I guess the trick there would be how to drive a larger LCD/OLED with a Daisy Seed or Teensy. Honestly, I don’t know enough about all of this stuff, so it’s definitely an exploration.

**Reply by djowel** [reply]
I recommended daisy or a teensy, only because these are already designed with audio in mind, although I haven't really used them :-) —I design my own. I recommend these dev boards, with the awareness of the difficulties of mixed signal PCB design. IMO, it's a magnitude more difficult than low-frequency analog/audio or digital alone. In my experience, I find dev boards that are not designed for audio too noisy, even at the lower 12 bits resolution.

**Reply by joulupukki** [reply]
Ok, I did some additional experimentation. I have a different ESP32 with a smaller OLED screen made by Heltec. It has a 32.768kHz external crystal on its dev board. Guess what? It detects the incoming frequency perfectly. I don't think the "Cheap Yellow Display" (ESP32-2432S028R) has an external crystal for precise ADC measurements. I'm trying to determine if that dev board can be modded to add an external crystal, but I'm guessing it's gonna be a no-go for that particular unit. Frustrating. I may look into the Teensy or Daisy Seed. I just know those seem to be quite a bit more expensive than some of the ESP32 options.

**Reply by joulupukki** [reply]
Alright, it turns out that what I was hitting is a bug in ESP32-WROOM with ESP-IDF. This frequency bug apparently doesn't happen in ESP32-S2 or ESP32-S3. But, I'm getting closer and Q DSP's pitch detector is FAST: https://youtu.be/lFYluL494ug

Latest code is in this branch but it's kind of a mess: https://github.com/joulupukki/esp32-chromatic-tuner/blob/boyd/switch-to-cyd/main/main.cpp

I've got a very lame hack/workaround to the ADC frequency issue on line 734.

**Reply by xconverge** [reply]
Looking good!

**Reply by djowel** [reply]
Great!
